syntax = "proto3";

package app;

service WordGame {
    rpc CreateGame(CreateGameRequest) returns (CreateGameResponse) {}
    rpc JoinGame(JoinGameRequest) returns (JoinGameResponse) {}
    rpc InitGame(InitGameRequest) returns (InitGameResponse) {}
    rpc GetPangram(GetPangramRequest) returns (GetPangramResponse) {}
    rpc SubmitWord(WordSubmissionRequest) returns (WordSubmissionResponse) {}
    rpc QueryGameStatus(GameStatusRequest) returns (GameStatusResponse) {}
    rpc EndGame(EndGameRequest) returns (EndGameResponse) {}
}

message CreateGameRequest {
    string username = 1
    string gameType = 2;
    string gameMode = 3;
}

message CreateGameResponse {
    bytes gameId = 1;
    string joinCode = 2;
}

message JoinGameRequest {
    string username = 1;
    string joinCode = 2;
}

message JoinGameResponse {
    int32 responseCode = 1;
    bytes gameId = 2;
}

message InitGameRequest {
    bytes gameId = 1;
}

message InitGameResponse {
    int32 responseCode = 1;
}

message GetPangramRequest {
    bytes gameId = 1;
}

message GetPangramResponse {
    string letters = 1;
}

message WordSubmissionRequest {
    bytes gameId = 1;
    string username = 2;
    string word = 3;
}

message WordSubmissionResponse {
    int32 score = 1;
    int32 total = 2;
    string message = 3;
}

message GameStatusRequest {
    bytes gameId = 1;
}

message GameStatusResponse {
    string status = 1;
}

message EndGameRequest {
    bytes gameId = 1;
}

message EndGameResponse {
    int32 response = 1;
}